File: jambeth/jambeth-platform/src/main/java/de/osthus/ambeth/platform/AmbethPlatformContext.java
package de.osthus.ambeth.platform;

import de.osthus.ambeth.collections.IList;
import de.osthus.ambeth.config.Properties;
import de.osthus.ambeth.database.IDatabaseProvider;
import de.osthus.ambeth.event.IEventQueue;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.ioc.IInitializingModule;
import de.osthus.ambeth.ioc.IModuleProvider;
import de.osthus.ambeth.ioc.IServiceContext;
import de.osthus.ambeth.ioc.factory.BeanContextFactory;
import de.osthus.ambeth.ioc.factory.IBeanContextFactory;
import de.osthus.ambeth.ioc.threadlocal.IThreadLocalCleanupController;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LoggerFactory;
import de.osthus.ambeth.merge.ILightweightTransaction;
import de.osthus.ambeth.objectcollector.IThreadLocalObjectCollector;
import de.osthus.ambeth.objectcollector.ThreadLocalObjectCollector;
import de.osthus.ambeth.persistence.IDatabase;
import de.osthus.ambeth.threading.IBackgroundWorkerDelegate;
import de.osthus.ambeth.threading.IBackgroundWorkerParamDelegate;
import de.osthus.ambeth.util.ModuleUtil;
import de.osthus.ambeth.util.ParamChecker;

public class AmbethPlatformContext implements IAmbethPlatformContext
{
	private class AmbethPlatformContextModule implements IInitializingModule
	{
		protected final AmbethPlatformContext apc;

		public AmbethPlatformContextModule(AmbethPlatformContext apc)
		{
			this.apc = apc;
		}

		@Override
		public void afterPropertiesSet(IBeanContextFactory beanContextFactory) throws Throwable
		{
			beanContextFactory.registerExternalBean(apc).autowireable(IAmbethPlatformContext.class);
		}
	}

	public static IAmbethPlatformContext create(Properties props, Class<?>[] providerModules, Class<?>[] frameworkModules, Class<?>[] bootstrapModules,
			IInitializingModule[] providerModuleInstances, final IInitializingModule[] frameworkModuleInstances,
			final IInitializingModule[] bootstrapModuleInstances)
	{
		ParamChecker.assertParamNotNull(props, "props");

		IServiceContext bootstrapContext = null;
		final AmbethPlatformContext apc = new AmbethPlatformContext();
		try
		{
			IInitializingModule[] providerModuleInstancesCopy = new IInitializingModule[providerModuleInstances.length + 1];
			System.arraycopy(providerModuleInstances, 0, providerModuleInstancesCopy, 0, providerModuleInstances.length);
			providerModuleInstancesCopy[providerModuleInstancesCopy.length - 1] = apc.new AmbethPlatformContextModule(apc);
			providerModuleInstances = providerModuleInstancesCopy;

			bootstrapContext = BeanContextFactory.createBootstrap(props, providerModules, providerModuleInstances);

			IList<IModuleProvider> moduleProviders = bootstrapContext.getImplementingObjects(IModuleProvider.class);
			for (int a = moduleProviders.size(); a-- > 0;)
			{
				IModuleProvider moduleProvider = moduleProviders.get(a);
				Class<?>[] mpFrameworkModules = moduleProvider.getFrameworkModules();
				Class<?>[] mpBootstrapModules = moduleProvider.getBootstrapModules();
				frameworkModules = ModuleUtil.mergeModules(mpFrameworkModules, frameworkModules);
				bootstrapModules = ModuleUtil.mergeModules(mpBootstrapModules, bootstrapModules);
			}
			IServiceContext frameworkBeanContext = bootstrapContext;

			if (frameworkModules.length > 0 || frameworkModuleInstances.length > 0)
			{
				frameworkBeanContext = bootstrapContext.createService("framework", new IBackgroundWorkerParamDelegate<IBeanContextFactory>()
				{
					@Override
					public void invoke(IBeanContextFactory childContextFactory)
					{
						for (int a = frameworkModuleInstances.length; a-- > 0;)
						{
							childContextFactory.registerExternalBean(frameworkModuleInstances[a]);
						}
					}
				}, frameworkModules);
			}

			ILightweightTransaction transaction = frameworkBeanContext.getService(ILightweightTransaction.class, false);
			if (transaction != null)
			{
				ILogger log = LoggerFactory.getLogger(AmbethPlatformContext.class, props);
				if (log.isInfoEnabled())
				{
					log.info("Starting initial database transaction to receive metadata for OR-Mappings...");
				}
				transaction.runInTransaction(new IBackgroundWorkerDelegate()
				{
					@Override
					public void invoke() throws Throwable
					{
						// Intended blank
					}
				});
				if (log.isInfoEnabled())
				{
					log.info("Initial database transaction processed successfully");
				}
			}
			IServiceContext applicationBeanContext = frameworkBeanContext;

			if (bootstrapModules.length > 0 || bootstrapModuleInstances.length > 0)
			{
				applicationBeanContext = frameworkBeanContext.createService("application", new IBackgroundWorkerParamDelegate<IBeanContextFactory>()
				{
					@Override
					public void invoke(IBeanContextFactory childContextFactory)
					{
						for (int a = bootstrapModuleInstances.length; a-- > 0;)
						{
							childContextFactory.registerExternalBean(bootstrapModuleInstances[a]);
						}
					}
				}, bootstrapModules);
			}
			apc.beanContext = applicationBeanContext;
			return apc;
		}
		catch (Throwable e)
		{
			if (bootstrapContext != null)
			{
				IThreadLocalCleanupController tlCleanupController = bootstrapContext.getService(IThreadLocalCleanupController.class);
				bootstrapContext.dispose();
				tlCleanupController.cleanupThreadLocal();
			}
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	protected IServiceContext beanContext;

	protected IEventQueue eventQueue;

	@Override
	public void dispose()
	{
		if (beanContext == null)
		{
			return;
		}
		IServiceContext rootContext = beanContext.getRoot();
		IThreadLocalObjectCollector tlObjectCollector = rootContext.getService(IThreadLocalObjectCollector.class);
		beanContext.getService(IThreadLocalCleanupController.class).cleanupThreadLocal();
		rootContext.dispose();
		beanContext = null;
		if (tlObjectCollector instanceof ThreadLocalObjectCollector)
		{
			((ThreadLocalObjectCollector) tlObjectCollector).clearThreadLocal();
			((ThreadLocalObjectCollector) tlObjectCollector).clearThreadLocals();
		}
	}

	@Override
	public IServiceContext getBeanContext()
	{
		return beanContext;
	}

	@Override
	public void clearThreadLocal()
	{
		IThreadLocalCleanupController threadLocalCleanupController = beanContext.getService(IThreadLocalCleanupController.class);
		threadLocalCleanupController.cleanupThreadLocal();
	}

	@Override
	public void afterBegin()
	{
		if (eventQueue != null)
		{
			eventQueue.enableEventQueue();
		}
	}

	@Override
	public void afterCommit()
	{
		afterEnd();
	}

	@Override
	public void afterRollback()
	{
		afterEnd();
	}

	protected void afterEnd()
	{
		try
		{
			if (eventQueue != null)
			{
				eventQueue.flushEventQueue();
			}
		}
		finally
		{
			try
			{
				IDatabaseProvider databaseProvider = beanContext.getService("databaseProvider", IDatabaseProvider.class, false);
				IDatabase database = databaseProvider != null ? databaseProvider.tryGetInstance() : null;
				if (database != null)
				{
					database.dispose();
				}
			}
			finally
			{
				clearThreadLocal();
			}
		}
	}
}

File: ambeth/Ambeth.Platform/ambeth/platform/AmbethPlatformContext.cs
using De.Osthus.Ambeth.Collections;
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Database;
using De.Osthus.Ambeth.Event;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Factory;
using De.Osthus.Ambeth.Ioc.Threadlocal;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Persistence;
using De.Osthus.Ambeth.Util;
using System;
using System.Collections.Generic;
namespace De.Osthus.Ambeth.Platform
{
    public class AmbethPlatformContext : IAmbethPlatformContext
    {
        private class AmbethPlatformContextModule : IInitializingModule
        {
            protected readonly AmbethPlatformContext apc;

            public AmbethPlatformContextModule(AmbethPlatformContext apc)
            {
                this.apc = apc;
            }

            public void AfterPropertiesSet(IBeanContextFactory beanContextFactory)
            {
                beanContextFactory.RegisterExternalBean(apc).Autowireable<IAmbethPlatformContext>();
            }
        }

        public static IAmbethPlatformContext Create(Properties props, Type[] providerModules, Type[] frameworkModules, Type[] bootstrapModules,
                IInitializingModule[] providerModuleInstances, IInitializingModule[] frameworkModuleInstances,
                IInitializingModule[] bootstrapModuleInstances)
        {
            ParamChecker.AssertParamNotNull(props, "props");

            IServiceContext bootstrapContext = null;
            AmbethPlatformContext apc = new AmbethPlatformContext();
            try
            {
                IInitializingModule[] providerModuleInstancesCopy = new IInitializingModule[providerModuleInstances.Length + 1];
                Array.Copy(providerModuleInstances, 0, providerModuleInstancesCopy, 0, providerModuleInstances.Length);
                providerModuleInstancesCopy[providerModuleInstancesCopy.Length - 1] = new AmbethPlatformContextModule(apc);
                providerModuleInstances = providerModuleInstancesCopy;

                bootstrapContext = BeanContextFactory.CreateBootstrap(props, providerModules, providerModuleInstances);

                IList<IModuleProvider> moduleProviders = bootstrapContext.GetImplementingObjects<IModuleProvider>();
                for (int a = moduleProviders.Count; a-- > 0; )
                {
                    IModuleProvider moduleProvider = moduleProviders[a];
                    Type[] mpFrameworkModules = moduleProvider.GetFrameworkModules();
                    Type[] mpBootstrapModules = moduleProvider.GetBootstrapModules();
                    frameworkModules = ModuleUtil.MergeModules(mpFrameworkModules, frameworkModules);
                    bootstrapModules = ModuleUtil.MergeModules(mpBootstrapModules, bootstrapModules);
                }
                IServiceContext frameworkBeanContext = bootstrapContext;

                if (frameworkModules.Length > 0 || frameworkModuleInstances.Length > 0)
                {
                    frameworkBeanContext = bootstrapContext.CreateService("framework", delegate(IBeanContextFactory childContextFactory)
                        {
                            for (int a = frameworkModuleInstances.Length; a-- > 0; )
                            {
                                childContextFactory.RegisterExternalBean(frameworkModuleInstances[a]);
                            }
                        }, frameworkModules);
                }

                ITransaction transaction = frameworkBeanContext.GetService<ITransaction>(false);
                if (transaction != null)
                {
                    ILogger log = LoggerFactory.GetLogger(typeof(AmbethPlatformContext), props);
                    if (log.InfoEnabled)
                    {
                        log.Info("Starting initial database transaction to receive metadata for OR-Mappings...");
                    }
                    transaction.ProcessAndCommit(delegate(ILinkedMap<Object, IDatabase> persistenceUnitToDatabaseMap)
                        {
                            // Intended blank
                        });
                    if (log.InfoEnabled)
                    {
                        log.Info("Initial database transaction processed successfully");
                    }
                }
                IServiceContext applicationBeanContext = frameworkBeanContext;

                if (bootstrapModules.Length > 0 || bootstrapModuleInstances.Length > 0)
                {
                    applicationBeanContext = frameworkBeanContext.CreateService("application", delegate(IBeanContextFactory childContextFactory)
                        {
                            for (int a = bootstrapModuleInstances.Length; a-- > 0; )
                            {
                                childContextFactory.RegisterExternalBean(bootstrapModuleInstances[a]);
                            }
                        }, bootstrapModules);
                }
                apc.beanContext = applicationBeanContext;
                return apc;
            }
            catch (Exception)
            {
                if (bootstrapContext != null)
                {
                    IThreadLocalCleanupController tlCleanupController = bootstrapContext.GetService<IThreadLocalCleanupController>();
                    bootstrapContext.Dispose();
                    tlCleanupController.CleanupThreadLocal();
                }
                throw;
            }
        }

        protected IServiceContext beanContext;

        protected IEventQueue eventQueue;

        public void Dispose()
        {
            if (beanContext == null)
            {
                return;
            }
            IServiceContext rootContext = beanContext.GetRoot();
            beanContext.GetService<IThreadLocalCleanupController>().CleanupThreadLocal();
            rootContext.Dispose();
            beanContext = null;
        }

        public IServiceContext GetBeanContext()
        {
            return beanContext;
        }

        public void ClearThreadLocal()
        {
            IThreadLocalCleanupController threadLocalCleanupController = beanContext.GetService<IThreadLocalCleanupController>();
            threadLocalCleanupController.CleanupThreadLocal();
        }

        public void AfterBegin()
        {
            if (eventQueue != null)
            {
                eventQueue.EnableEventQueue();
            }
        }

        public void AfterCommit()
        {
            AfterEnd();
        }

        public void AfterRollback()
        {
            AfterEnd();
        }

        protected void AfterEnd()
        {
            try
            {
                if (eventQueue != null)
                {
                    eventQueue.FlushEventQueue();
                }
            }
            finally
            {
                try
                {
                    IDatabaseProvider databaseProvider = beanContext.GetService<IDatabaseProvider>("databaseProvider", false);
                    IDatabase database = databaseProvider != null ? databaseProvider.TryGetInstance() : null;
                    if (database != null)
                    {
                        database.Dispose();
                    }
                }
                finally
                {
                    ClearThreadLocal();
                }
            }
        }
    }
}